### PR TITLE
LibJS: Add initial implementation for SharedArrayBuffer

### PR DESCRIPTION
**Pull request by shannonbooth**
None of the actual sharing is implemented yet, but this is enough for most basic functionality.

Diff Tests:
    +260 ✅    -262 ❌   +2 💀

### CODE REVIEW

File: Userland/Libraries/LibJS/Runtime/SharedArrayBufferPrototype.cpp [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+}
+
+ThrowCompletionOr<void> SharedArrayBufferPrototype::initialize(Realm& realm)
+{
+    auto& vm = this->vm();
+
+    MUST_OR_THROW_OOM(Base::initialize(realm));
+    u8 attr = Attribute::Writable | Attribute::Configurable;
+    define_native_function(realm, vm.names.slice, slice, 2, attr);
+    define_native_accessor(realm, vm.names.byteLength, byte_length_getter, {}, Attribute::Configurable);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/SharedArrayBufferPrototype.cpp:27]: swap order

File: Userland/Libraries/LibJS/Runtime/SharedArrayBufferPrototype.cpp [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    auto array_buffer_object = TRY(typed_this_value(vm));
+
+    // 3. If IsSharedArrayBuffer(O) is false, throw a TypeError exception.
+    // FIXME: Check for shared buffer
+
+    // 4. Let len be O.[[ArrayBufferByteLength]].
+    auto length = array_buffer_object->byte_length();
+
+    // 5. Let relativeStart be ? ToIntegerOrInfinity(start).
+    auto relative_start = TRY(vm.argument(0).to_integer_or_infinity(vm));
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/SharedArrayBufferPrototype.cpp:66]: Please `auto start = vm.argument(0);` once at the beginning for readability (same for `end`)

File: Userland/Libraries/LibJS/Runtime/SharedArrayBufferPrototype.cpp [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+    // 14. Let ctor be ? SpeciesConstructor(O, %ArrayBuffer%).
+    auto* constructor = TRY(species_constructor(vm, array_buffer_object, realm.intrinsics().shared_array_buffer_constructor()));
+
+    // 15. Let new be ? Construct(ctor, « 𝔽(newLen) »).
+    auto new_array_buffer = TRY(construct(vm, *constructor, Value(new_length)));
+
+    // 16. Perform ? RequireInternalSlot(new, [[ArrayBufferData]]).
+    if (!is<ArrayBuffer>(new_array_buffer.ptr()))
+        return vm.throw_completion<TypeError>(ErrorType::SpeciesConstructorDidNotCreate, "an SharedArrayBuffer");
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/SharedArrayBufferPrototype.cpp:104]: doesn't have to be a SAB, can also be a plain ArrayBuffer

File: Userland/Libraries/LibJS/Runtime/SharedArrayBufferPrototype.cpp [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    auto* new_array_buffer_object = static_cast<ArrayBuffer*>(new_array_buffer.ptr());
+
+    // 17. If IsSharedArrayBuffer(new) is true, throw a TypeError exception.
+    // FIXME: Check for shared buffer
+
+    // FIXME: 18. If new.[[ArrayBufferData]] is O.[[ArrayBufferData]], throw a TypeError exception.
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/SharedArrayBufferPrototype.cpp:110]: This is a simple pointer compare, no reason to not implement it :)

File: Userland/Libraries/LibJS/Tests/builtins/SharedArrayBuffer/SharedArrayBuffer.prototype.byteLength.js [hunk 5]
```diff
@@ -0,0 +1,6 @@
+test("basic functionality", () => {
```
Comment (linusg) [inline Userland/Libraries/LibJS/Tests/builtins/SharedArrayBuffer/SharedArrayBuffer.prototype.byteLength.js:1]: Can we have a test for slice() please?

**Review by linusg — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp [hunk 6]
```diff
@@ -229,4 +229,21 @@ ThrowCompletionOr<ArrayBuffer*> array_buffer_copy_and_detach(VM& vm, ArrayBuffer
     return new_buffer;
 }
 
+// 25.2.1.1 AllocateSharedArrayBuffer ( constructor, byteLength ), https://tc39.es/ecma262/#sec-allocatesharedarraybuffer
+ThrowCompletionOr<ArrayBuffer*> allocate_shared_array_buffer(VM& vm, FunctionObject& constructor, size_t byte_length)
+{
+    // 1. Let obj be ? OrdinaryCreateFromConstructor(constructor, "%SharedArrayBuffer.prototype%", « [[ArrayBufferData]], [[ArrayBufferByteLength]] »).
+    auto obj = TRY(ordinary_create_from_constructor<ArrayBuffer>(vm, constructor, &Intrinsics::shared_array_buffer_prototype, nullptr));
+
+    // FIXME: 2. Let block be ? CreateSharedByteDataBlock(byteLength).
```
[the remark this answers is not in the document]
Comment (shannonbooth) [reply Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp:238]: I'm calling the non-shared array buffer IO here, probably I can just improve the FIXME?

File: Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp [hunk 7]
```diff
@@ -229,4 +229,21 @@ ThrowCompletionOr<ArrayBuffer*> array_buffer_copy_and_detach(VM& vm, ArrayBuffer
     return new_buffer;
 }
 
+// 25.2.1.1 AllocateSharedArrayBuffer ( constructor, byteLength ), https://tc39.es/ecma262/#sec-allocatesharedarraybuffer
+ThrowCompletionOr<ArrayBuffer*> allocate_shared_array_buffer(VM& vm, FunctionObject& constructor, size_t byte_length)
```
[the remark this answers is not in the document]
Comment (shannonbooth) [reply Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp:233]: Ah, yup got this from copy paste. I will fix up the place that I took it from while I'm at it :)

File: Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (linusg) [reply Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp:238]: ah, missed that. Let's leave it as is then.

File: Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (shannonbooth) [reply Userland/Libraries/LibJS/Runtime/ArrayBuffer.cpp]: Done! I'll make a follow up MR at some stage to move over the other ones away from raw pointers.

File: Userland/Libraries/LibJS/Runtime/SharedArrayBufferPrototype.cpp (hunk 2 above)
Comment (shannonbooth) [reply Userland/Libraries/LibJS/Runtime/SharedArrayBufferPrototype.cpp]: Yup, much nicer. Also included a comment for same change on the ArrayBufferPrototype.

File: Userland/Libraries/LibJS/Runtime/SharedArrayBufferPrototype.cpp (hunk 4 above)
Comment (shannonbooth) [reply Userland/Libraries/LibJS/Runtime/SharedArrayBufferPrototype.cpp]: Thanks! Confirmed it fixes that issue :^)

File: Userland/Libraries/LibJS/Tests/builtins/SharedArrayBuffer/SharedArrayBuffer.prototype.byteLength.js (hunk 5 above)
Comment (shannonbooth) [reply Userland/Libraries/LibJS/Tests/builtins/SharedArrayBuffer/SharedArrayBuffer.prototype.byteLength.js:1]: I've copy pasted over the tests from ArrayBuffer. Doesn't feel that great. Let me know if you think I should extract these into a common file (not sure what I would call it) for these tests that can be shared between ArrayBuffer and SharedArrayBuffer

File: Userland/Libraries/LibJS/Runtime/SharedArrayBufferPrototype.cpp (hunk 3 above)
Comment (shannonbooth) [reply Userland/Libraries/LibJS/Runtime/SharedArrayBufferPrototype.cpp]: Good point, changed the wording

**Pull request merged**